crates/noirc_evaluator/src/brillig/mod.rs

```diff
@@ -46,15 +49,46 @@ impl std::ops::Index<FunctionId> for Brillig {
 }
 
 impl Ssa {
-    /// Generate compilation artifacts for brillig functions
+    /// Compile to brillig brillig functions and ACIR functions reachable from them
     pub(crate) fn to_brillig(&self) -> Brillig {
-        // Collect all of the brillig functions
-        let brillig_functions =
-            self.functions.values().filter(|func| func.runtime() == RuntimeType::Brillig);
+        // Collect all the function ids that are reachable from brillig
+        // That means all the functions marked as brillig and ACIR functions called by them
+        let mut reachable_function_ids: HashSet<FunctionId> = HashSet::new();
+
+        // Initialize the queue with all the functions marked as brillig
+        let mut reachability_queue: Vec<FunctionId> = self
+            .functions
+            .iter()
+            .filter_map(
+                |(id, func)| {
+                    if func.runtime() == RuntimeType::Brillig {
+                        Some(*id)
+                    } else {
+                        None
+                    }
+                },
+            )
+            .collect();
+
+        while !reachability_queue.is_empty() {
+            let func = &self.functions[&reachability_queue.pop().unwrap()];
+            reachable_function_ids.insert(func.id());
+            for (_, value) in func.dfg.values_iter() {
+                // All reachable functions appear as literals after defunctionalization of the SSA
+                if let Value::Function(function_id) = value {
```

Comment:
This section has quite a bit of rightwards drift, some pseudocode that reduces it would look like:

```rust
while let Some(func_id) = reachability_queue.pop() {
    let func = &self.functions[&func_id];
    reachable_function_ids.insert(func.id());

    for (_, value) in func.dfg.values_iter() {
        // All reachable functions appear as literals after defunctionalization of the SSA
        // If value is not a function or if it's already reachable or in queue, skip
        let function_id = match value {
            Value::Function(function_id) => function_id,
            _ => continue,
        };
        if reachable_function_ids.contains(function_id) || reachability_queue.contains(function_id) {
            continue;
        }
        
        reachability_queue.push(*function_id);
    }
}
```